### PR TITLE
Raise max mapgen limit constant to align with MapBlock

### PR DESCRIPTION
**Pull request by sfan5**
basic unittests included

## To do

This PR is Ready for Review.

### CODE REVIEW

File: src/map.cpp [hunk 1]
```diff
@@ -1444,11 +1444,7 @@ MapSector *ServerMap::createSector(v2s16 p2d)
 	/*
 		Do not create over max mapgen limit
 	*/
-	const s16 max_limit_bp = MAX_MAP_GENERATION_LIMIT / MAP_BLOCKSIZE;
-	if (p2d.X < -max_limit_bp ||
-			p2d.X >  max_limit_bp ||
-			p2d.Y < -max_limit_bp ||
-			p2d.Y >  max_limit_bp)
+	if (blockpos_over_max_limit(v3s16(p2d.X, 0, p2d.Y)))
```
Comment (erlehmann) [inline src/map.cpp:1447]: What exactly is the purpose of this change? It looks like it fixes a bug, but it is unclear to me what the problem was here before and why this is part of this PR.

The code reuses blockpos_over_max_limit() with the Y component of a v3s16 set to 0 and the Z component of the v3s16 set to the Y component of a v2s16. Is this intentional?

File: src/mapblock.h [hunk 2]
```diff
@@ -601,7 +601,7 @@ typedef std::vector<MapBlock*> MapBlockVect;
 
 inline bool objectpos_over_limit(v3f p)
 {
-	const float max_limit_bs = MAX_MAP_GENERATION_LIMIT * BS;
+	const float max_limit_bs = (MAX_MAP_GENERATION_LIMIT + 0.5f) * BS;
```
Comment (erlehmann) [inline src/mapblock.h:604]: What is the purpose of this change? Is this to align legal entity and player positions with the edge of the outermost mapblock? If so, why is it part of this PR? It seems like a separate issue to me.

**Review by erlehmann — COMMENTED** [review]
This looks like a step in the right direction, but I have questions.

**Comment by erlehmann** [comment]
Some mapgen code seems to hardcode 31000 instead of MAX_MAP_GENERATION_LIMIT, by mistake IMO.

I think this PR is a good opportunity to fix that, so 31007 is used everywhere from now on.

You should be able to find them using `git grep -W 31000 src` in the Minetest repository.

File: src/map.cpp (hunk 1 above)
Comment (sfan5) [reply src/map.cpp:1447]: Just less code for the same thing, no behavioral change.

> Y component of a v3s16 set to 0 and the Z component of the v3s16 set to the Y component of a v2s16. Is this intentional?

Yes. The X and Z components of a block positions specify the sector it is in.

File: src/mapblock.h (hunk 2 above)
Comment (sfan5) [reply src/mapblock.h:604]: > Is this to align legal entity and player positions with the edge of the outermost mapblock?

Yes. Creating a separate PR for this is unnecessary IMO.

**Comment by sfan5** [comment]
For fields like `y_min` or `y_max` the default of (-)31000 is documented part of the API so changing them could run into problems. Unlike the mapgen limit the impact of these is also minimal.

The `dungeon_ymin` and `floatland_ywater` settings *could* be changed.
But this would also make no difference since due to the an extra border reserved by the emerge code the effective boundary remains at 30991 (which is below 31000).

File: src/unittest/test_map.cpp [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+	UASSERT(objectpos_over_limit(v3f(-limit_times_bs+BS/2)) == false);
+	UASSERT(objectpos_over_limit(v3f(-limit_times_bs)) == false);
+	UASSERT(objectpos_over_limit(v3f(-limit_times_bs-BS/2)) == false);
+	UASSERT(objectpos_over_limit(v3f(-limit_times_bs-BS)) == true);
+
+	// blockpos_over_max_limit
+	s16 limit_block = MAX_MAP_GENERATION_LIMIT / MAP_BLOCKSIZE;
+	UASSERT(blockpos_over_max_limit(v3s16(limit_block)) == false);
+	UASSERT(blockpos_over_max_limit(v3s16(limit_block+1)) == true);
```
Comment (x2048) [inline src/unittest/test_map.cpp:65]: It might make sense to check each component separately, as this line tests only .X

For example:
```
	UASSERT(blockpos_over_max_limit(v3s16(limit_block+1, 0, 0)) == true);
	UASSERT(blockpos_over_max_limit(v3s16(0, limit_block+1, 0)) == true);
	UASSERT(blockpos_over_max_limit(v3s16(0, 0, limit_block+1)) == true);
```
And for the negative
```
	UASSERT(blockpos_over_max_limit(v3s16(-limit_block-1, 0, 0)) == true);
	UASSERT(blockpos_over_max_limit(v3s16(0, -limit_block-1, 0)) == true);
	UASSERT(blockpos_over_max_limit(v3s16(0, 0, -limit_block-1)) == true);
```

(I tried to propose changes but github did not want to let me do it.)

Comment (x2048) [reply src/unittest/test_map.cpp:65]: Same for lines 55 and 60, just noticed.

Comment (sfan5) [reply src/unittest/test_map.cpp:65]: > as this line tests only .X

it tests `X`, `Y` and `Z` actually. This isn't comprehensive either of course but I think it's probably good enough.

Comment (appgurueu) [reply src/unittest/test_map.cpp:65]: Yeah, sfan is right, see the [Irrlicht docs](https://irrlicht.sourceforge.io/docu/classirr_1_1core_1_1vector3d.html#a08093c70044d66a5c3cdec2522b868f2). Why is this not resolved yet?

**Comment by Df458** [comment]
> You should be able to find the code by executing git grep -W 31000 src in the repository.

And indeed, with a search we find a few hardcoded spots not mentioned previously:
- minetest.conf.example, lines 1830-1834: The mapgen limit is listed here as 31000, should be updated to reflect the new limit
- chat.lua, line 527 (teleport_to_pos): Limit is hardcoded to 31000. Do we care?


Biome definitions also have a default `min_pos`/`max_pos` hardcoded to +-31000, but this is documented so I assume we'll treat them the same way as `y_min` and `y_max`

**Comment by Df458** [comment]
Anyway, I looked at the code and what I saw made sense. I still have misgivings about things outside this change we could've overlooked, but I'd say it passes review. However, without a bug repro or visible feature that I can easily test I'm not prepared to approve.

@erlehmann Since you brought up the issue mentioning incorrect assumptions in mod code, do you have any specific examples we could use to verify this?

**Comment by sfan5** [comment]
Turns out I missed a few more, thanks for pointing that out.

**Comment by erlehmann** [comment]
> @erlehmann Since you brought up the issue mentioning incorrect assumptions in mod code, do you have any specific examples we could use to verify this?

There were no “incorrect assumptions in mod code”, until  https://github.com/minetest/minetest/pull/11770 erroneously clamped the output of `minetest.find_nodes_in_area()` to MAX_MAPGEN_LIMIT to fix an integer overflow This made it so that MInetest can place nodes that it can not find itself anymore.

Placing nodes and then finding them with `mintest.find_nodes_in_area()` later isn't exactly a *rare* pattern in Minetest.

It happens in Mineclonia in placement callbacks for a structure, in village generation, when using ice walker enchantment, when placing wells and fossils and numerous other stuff in the map generator, and, crucially, when removing water, stone, sand and dirt from the Nether dimension.


https://git.minetest.land/Mineclonia/Mineclonia/src/branch/master/mods/MISC/mcl_selftests/init.lua

**Comment by Df458** [comment]
That list of interaction is good, yes. I'll try to give this a more thorough test using Mineclonia in the next few hours

**Comment by erlehmann** [comment]
> That list of interaction is good, yes. I'll try to give this a more thorough test using Mineclonia in the next few hours

The most obvious thing beside the unit test I can think of right now is placing a structure, then not being able to find all the nodes that were placed. This was not a problem before https://github.com/minetest/minetest/pull/11770 and I would like it to not be a problem in the future.

Please file any bugs you find in the bug tracker. Thank you!

Btw, a hilarious failure mode of having an area which exists, but is considered “out of bounds” is this bug https://github.com/minetest/minetest/issues/11839 where a TNT mod fills the disk because the entity it creates is removed and recreated a lot.

**Review by Df458 — APPROVED** [review]
Did some testing with the code from Mineclonia, I think this is fine to merge. I did notice an overflow in `find_nodes_in_area`'s argument parsing, but it's unrelated to this change and already existed so it shouldn't block.
